### PR TITLE
LOADKEY Function: Add management of persistent handle

### PR DESCRIPTION
**Pull request by GaelGuegan**
Here the patch of the issue #10 .

### CODE REVIEW

File: include/tpm2-tss-engine.h [hunk 1]
```diff
@@ -34,10 +34,15 @@
 #include <openssl/engine.h>
 #include <tss2/tss2_tpm2_types.h>
 
+typedef enum { blob, handle } KEY_TYPE ;
```
Comment (AndreasFuchsSIT) [inline include/tpm2-tss-engine.h]: Please namespace the enum members, preferably with the enum typename, i.e. KEY_TYPE_BLOB

Comment (GaelGuegan) [reply include/tpm2-tss-engine.h]: Okay can I use tpm_tools scripts in my test, in order to generate a persistent handle ?

**Comment by AndreasFuchsSIT** [comment]
Sure, that would even be good for people to look up.
I can also help with the .travis.yml scripts for installing the tpm2-tools on the CI-server

**Comment by GaelGuegan** [comment]
Hey @AndreasFuchsSIT, I try several things but I have still errors because of package dependencies with tpm-tools. I try to add libcurl3-dev & libtool to travis.yml, without success ...

**Comment by AndreasFuchsSIT** [comment]
maybe try requiring libcurl4-openssl-dev isntead of libcurl3-dev

**Comment by GaelGuegan** [comment]
Okay, I have tried what you suggest without success. A new error appear and I changed libgcrypt20-dev to libgcrypt-dev, but now it is the TSS compilation who does not work ... :'(
I am loosing hope.

**Comment by AndreasFuchsSIT** [comment]
Yes, tpm2-tss requires libgcrypt20-dev...
I guess is could be cloned and installed manually, similar to how tpm2-tss is cloned and installed as well.
Then this might just work...

If all fails, we'll repleace the tpm2-tools with some special binary using esapi directly. But let's see first...
________________________________

File: src/tpm2-tss-engine.c [hunk 2]
```diff
@@ -165,16 +166,50 @@ loadkey(ENGINE *e, const char *key_id, UI_METHOD *ui, void *cb_data)
...
 
 	DBG("Loading private key %s\n", key_id);
-
-    if (!tpm2tss_tpm2data_read(key_id, &tpm2Data)) {
+    tpm2Data = OPENSSL_malloc(sizeof(*tpm2Data));
+    if (tpm2Data == NULL) {
+      ERR(tpm2tss_tpm2data_read, ERR_R_MALLOC_FAILURE);
+      goto error;
+    }
+    if (strncmp(key_id, "0x", 2) == 0) {
```
Comment (dwmw2) [inline src/tpm2-tss-engine.c]: Please let's standardise a URI format for this, perhaps based on https://tools.ietf.org/html/draft-mavrogiannopoulos-tpmuri-01

File: src/tpm2-tss-engine-rsa.c [hunk 3]
```diff
@@ -246,8 +261,19 @@ rsa_priv_dec(int flen,
...
-    r = init_tpm_key(&ectx, &keyHandle, tpm2Data);
-    ERRchktss(rsa_priv_dec, r, goto out);
+    if (tpm2Data->privatetype == handle) {
+      r = Esys_Initialize(&ectx, NULL, NULL);
+      ERRchktss(rsa_priv_dec, r, goto out);
+      r = Esys_Startup(ectx, TPM2_SU_CLEAR);
+      ERRchktss(rsa_priv_dec, r, goto out);
+      r = Esys_TR_FromTPMPublic(ectx, tpm2Data->handle, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE, &keyHandle);
+      ERRchktss(rsa_priv_dec, r, goto out);
+      r = Esys_TR_SetAuth(ectx, ESYS_TR_RH_OWNER, &tpm2Data->userauth);
```
Comment (dwmw2) [inline src/tpm2-tss-engine-rsa.c]: You want to do a ReadPublic call here and check the NODA flag, then if the NODA flag isn't set, use a UI to prompt for a password if you don't already have one.

See http://git.infradead.org/users/dwmw2/openconnect.git/blob/tpm2:/gnutls_tpm2_esys.c#l277 for example.

File: src/tpm2-tss-engine.c (hunk 2 above)
Comment (AndreasFuchsSIT) [reply src/tpm2-tss-engine.c]: That thing is outdated for TPM2...
No uuids anymore and "storage" has different means of system vs user...
I'd wanna go with 0x or 0x81 for the moment, until we have figured out a tpm2uri scheme or adapted tpmuri...

File: src/tpm2-tss-engine-rsa.c (hunk 3 above)
Comment (GaelGuegan) [reply src/tpm2-tss-engine-rsa.c]: Okay I corrected it

**Comment by dwmw2** [comment]
For the password handling part, please do some dev testing of this manually with the engine being used from OpenConnect. You don't need a VPN; just point it at any HTTPS server and you'll test the password prompting.

For the builds with GnuTLS, it will automatically prompt the user as needed for the owner hierarchy password or the parent key password as appropriate, and for a password for the key itself.

If NODA is set on a key or a persistent parent (and always for a hierarchy password), it'll try the empty password and then prompt the user if that fails.

I have done my testing with all variants of hierarchy password set/unset, parent key password set/unset, actual key password set/unset. Most of my tests fail with the ENGINE while they work in the GnuTLS build.

**Comment by GaelGuegan** [comment]
I have revert to one of first commit.
If you want to correct other issue please make a pull request.

I cannot correct everyting ... sorry

**Comment by AndreasFuchsSIT** [comment]
That's fine, I'll take a stab.
Thanks for the effort up to here !

**Comment by AndreasFuchsSIT** [comment]
I took your stuff an refactored a bit. Please see #33 and reviews are welcome.
Closing this PR for the moment.

**Pull request closed without merging**